src/tensorflow.cc

```diff
@@ -1332,62 +1340,264 @@ AutoCompleteHelper::FixBatchingSupport()
+    TRITONTF_IO* io = item->io_;
+    if (!strcmp(name, io->name_)) {
+      return io;
+    }
+  }
+
+  return nullptr;
+}
+
+TRITONTF_IO*
+FindModelIOByName(const char* name, const TRITONTF_IOList* list)
+{
+  if (name == nullptr) {
+    return nullptr;
+  }
+
+  while (list) {
+    TRITONTF_IO* io = list->io_;
+    if (!strcmp(name, io->name_)) {
+      return io;
+    }
+    list = list->next_;
+  }
+
+  return nullptr;
+}
+
+
 TRITONSERVER_Error*
-AutoCompleteHelper::FixIOConfig(
-    const TRITONTF_IOList* reference_list, const char* key)
+AutoCompleteHelper::FillMissingValues(
+    TRITONTF_IO* io, triton::common::TritonJson::Value& config)
 {
-  // Replace I/O even if inputs / outputs are specified in config.
-  triton::common::TritonJson::Value ios;
-  bool found_ios = model_state_->ModelConfig().Find(key, &ios);
+  triton::common::TritonJson::Value tmp(
+      model_state_->ModelConfig(),
+      triton::common::TritonJson::ValueType::OBJECT);
+
+  bool found_name = config.Find("name", &tmp);
+  if (!found_name) {
+    triton::common::TritonJson::Value name;
+    RETURN_IF_ERROR(config.AddString("name", io->name_));
+  }
+
+  bool found_config_data_type = config.Find("data_type", &tmp);
+  std::string data_type_str;
+  tmp.AsString(&data_type_str);
+  bool should_auto_complete_data_type =
+      !found_config_data_type || DataTypeIsInvalid(data_type_str);
+  if (should_auto_complete_data_type) {
+    config.SetString("data_type", ConvertToModelConfigString(io->data_type_));
+  }
+
+  bool found_dims = config.Find("dims", &tmp);
+  bool should_auto_complete_dims = !found_dims || (tmp.ArraySize() == 0);
+  if (should_auto_complete_dims) {
+    RETURN_ERROR_IF_TRUE(
+        io->shape_->rank_ == 0, TRITONSERVER_ERROR_INVALID_ARG,
+        std::string(
+            "Unable to autofill for '" + model_state_->Name() +
+            "': the rank of model tensor '" + io->name_ +
+            "' is 0 and dimensions are not defined"));
+    // The model signature supports batching then the first
+    // dimension is -1 and should not appear in the model
+    // configuration 'dims' that we are creating.
+    triton::common::TritonJson::Value dims(
+        model_state_->ModelConfig(),
+        triton::common::TritonJson::ValueType::ARRAY);
+
+    for (size_t i = (model_support_batching_ ? 1 : 0); i < io->shape_->rank_;
+         ++i) {
+      RETURN_IF_ERROR(dims.AppendInt(io->shape_->dims_[i]));
+    }
+    if (dims.ArraySize() == 0) {
+      RETURN_IF_ERROR(dims.AppendInt(1));
+      triton::common::TritonJson::Value reshape(
+          model_state_->ModelConfig(),
+          triton::common::TritonJson::ValueType::OBJECT);
+      triton::common::TritonJson::Value reshape_dims(
+          model_state_->ModelConfig(),
+          triton::common::TritonJson::ValueType::ARRAY);
+      RETURN_IF_ERROR(reshape.Add("shape", std::move(reshape_dims)));
+      RETURN_IF_ERROR(config.Add("reshape", std::move(reshape)));
+    }
+
+    if (found_dims) {
```

Comment:
This doesn't look right. Why are we removing dims if found_dims == true?

Reply:
Ahh.. This is within if block with  `should_auto_complete_dims`. At the end of the if block just comment 

//  end of should_auto_complete_dims

Reply:
Correct, there are times where we may have dims defined but the array is empty. We must check for the presence of this field in the JSON otherwise blindly adding the autocompleted dims creates a duplicate field.

Reply:
Comment added